pkg/tkctl/cmd/dump/dump.go

```diff
@@ -0,0 +1,518 @@
+	cmd.Flags().DurationVar(&o.since, "since", 3600000000000, "Return logs newer than a relative duration like 1m, or 3h.")
+	cmd.Flags().Int64Var(&o.byteReadLimit, "byteReadLimit", 500000, "The maximum number of bytes dump log.")
+	cmd.MarkFlagRequired("path")
+	return cmd
+}
+
+func (o *dumpInfoOptions) Complete(tkcContext *config.TkcContext, cmd *cobra.Command, args []string) error {
+	clientConfig, err := tkcContext.ToTkcClientConfig()
+	if err != nil {
+		return err
+	}
+
+	if tidbClusterName, ok := clientConfig.TidbClusterName(); ok {
+		o.tidbClusterName = tidbClusterName
+	} else {
+		return cmdutil.UsageErrorf(cmd, dumpUsage)
+	}
+
+	namespace, _, err := clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+	o.namespace = namespace
+
+	restConfig, err := clientConfig.RestConfig()
+	if err != nil {
+		return err
+	}
+
+	tcCli, err := versioned.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+	o.tcCli = tcCli
+
+	kubeCli, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+	o.kubeCli = kubeCli
+
+	internalCli, err := internalclientset.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+	o.internalCli = internalCli
+
+	dynamicCli, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+	o.dynamicCli = dynamicCli
+
+	return nil
+}
+
+func (o *dumpInfoOptions) Run() error {
+	if err := createPathIfNotExist(o.logPath); err != nil {
+		return err
+	}
+
+	tc, err := o.tcCli.PingcapV1alpha1().
+		TidbClusters(o.namespace).
+		Get(o.tidbClusterName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// dump tidb cluster object information.
+	if err := NewTiDBClusterDescriber(o.internalCli, o.dynamicCli, tc).Dump(o.logPath); err != nil {
+		return err
+	}
+
+	// dump stateful information by a particular tidb cluster.
+	if err := NewTiDBClusterStatefulDescriber(*o.internalCli, tc).Dump(o.logPath); err != nil {
+		return err
+	}
+
+	podList, err := o.kubeCli.CoreV1().Pods(o.namespace).List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s,%s in (%s)", label.InstanceLabelKey, tc.Name, label.ComponentLabelKey,
+			strings.Join([]string{label.TiDBLabelVal, label.TiKVLabelVal, label.PDLabelVal}, ",")),
+	})
+	if err != nil {
+		return err
+	}
+
+	// dump detail information and logs of pods.
+	for _, pod := range podList.Items {
+		if err := NewPodDescriber(o.kubeCli, o.internalCli, pod, int64(o.since.Seconds()), o.byteReadLimit).Dump(o.logPath); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// tidbClusterDescriber generates information about a tidbclusters object.
+type tidbClusterDescriber struct {
+	internalCli *internalclientset.Clientset
+	dynamicCli  dynamic.Interface
+
+	mapping *meta.RESTMapping
+	tc      *v1alpha1.TidbCluster
+}
+
+// NewTiDBClusterDummper returns a tidbClusterDumper.
+func NewTiDBClusterDescriber(internalCli *internalclientset.Clientset, dynamicCli dynamic.Interface, tc *v1alpha1.TidbCluster) *tidbClusterDescriber {
```

Comment:
Just print the `TidbCluster` to `yaml` or `json` to a file is ok. No need `describe` them.

Reply:
Done